File: kotlin/src/13.kt
import kotlin.time.measureTimedValue

fun main() {
    val clawMachines = mutableListOf<List<String>>()
    var ls = readInput()
    while (ls.isNotEmpty()) {
        clawMachines.add(ls)
        ls = readInput()
    }

    fun parseButton(line: String): Pair<Long, Long> {
        val pattern = Regex("^Button [A-Z]: X\\+(\\d+), Y\\+(\\d+)$")
        val match = pattern.find(line)
        if (match != null) {
            val (_, x, y) = match.groupValues
            return Pair(x.toLong(), y.toLong())
        }
        return Pair(0, 0)
    }

    fun parsePrize(line: String): Pair<Long, Long> {
        val pattern = Regex("^Prize: X=(\\d+), Y=(\\d+)$")
        val match = pattern.find(line)
        if (match != null) {
            val (_, x, y) = match.groupValues
            return Pair(x.toLong(), y.toLong())
        }
        return Pair(0, 0)
    }

    fun nbTokens(buttonA: Pair<Long, Long>, buttonB: Pair<Long, Long>, prize: Pair<Long, Long>): Long {
        val det = buttonA.first * buttonB.second - buttonA.second * buttonB.first
        if (det == 0L)
            return 0

        val x = buttonB.second * prize.first - buttonB.first * prize.second
        if (x % det != 0L)
            return 0
        val a = x / det

        val y = - buttonA.second * prize.first + buttonA.first * prize.second
        if (y % det != 0L)
            return 0
        val b = y / det

        if (a < 0 || b < 0)
            return 0

        return 3 * a + b
    }

    fun part1(): Long {
        var ans = 0L

        for (clawMachine in clawMachines) {
            val buttonA = parseButton(clawMachine[0])
            val buttonB = parseButton(clawMachine[1])
            val prize = parsePrize(clawMachine[2])

            ans += nbTokens(buttonA, buttonB, prize)
        }

        return ans
    }

    fun part2(): Long {
        var ans = 0L

        for (clawMachine in clawMachines) {
            val buttonA = parseButton(clawMachine[0])
            val buttonB = parseButton(clawMachine[1])
            val prize = parsePrize(clawMachine[2]) + Pair(10000000000000, 10000000000000)

            ans += nbTokens(buttonA, buttonB, prize)
        }

        return ans
    }

    println(measureTimedValue { part1() }.let { "${it.value} (${it.duration})" })
    println(measureTimedValue { part2() }.let { "${it.value} (${it.duration})" })
}

File: 13.py
from aoc import *

DAY = 13

s = get_input(DAY)
# s = get_example(DAY)

import collections
import math
import re

ls = [[]]
for line in s.splitlines():
    if line == "":
        ls.append([])
    else:
        ls[-1].append(line)


def parse_button(line):
    pattern = r"^Button [A-Z]: X\+(\d+), Y\+(\d+)$"
    match = re.search(pattern, line)
    return int(match.group(1)), int(match.group(2))


def parse_prize(line):
    pattern = r"^Prize: X=(\d+), Y=(\d+)$"
    match = re.search(pattern, line)
    return int(match.group(1)), int(match.group(2))


def nb_tokens(button_a, button_b, prize):
    det = button_a[0] * button_b[1] - button_a[1] * button_b[0]
    if det == 0:
        return 0

    x = button_b[1] * prize[0] - button_b[0] * prize[1]
    if x % det != 0:
        return 0
    a = x // det

    y = - button_a[1] * prize[0] + button_a[0] * prize[1]
    if y % det != 0:
        return 0
    b = y // det

    if a < 0 or b < 0:
        return 0

    return 3 * a + b


def part1():
    ans = 0

    for l in ls:
        button_a = parse_button(l[0])
        button_b = parse_button(l[1])
        prize = parse_prize(l[2])

        ans += nb_tokens(button_a, button_b, prize)

    return ans


def part2():
    ans = 0

    for l in ls:
        button_a = parse_button(l[0])
        button_b = parse_button(l[1])
        prize = tuple(map(lambda x: x + 10000000000000, parse_prize(l[2])))

        ans += nb_tokens(button_a, button_b, prize)

    return ans


# submit(DAY, part1(), 1)
# submit(DAY, part2(), 2)
